File: src/alpaka/bin/EventProcessor.cc
#include <cmath>
#include <exception>
#include <filesystem>
#include <string>
#include <vector>

#include "Framework/ESPluginFactory.h"
#include "Framework/WaitingTask.h"
#include "Framework/WaitingTaskHolder.h"

#include "EventProcessor.h"

namespace edm {
  EventProcessor::EventProcessor(int warmupEvents,
                                 int maxEvents,
                                 int runForMinutes,
                                 int numberOfStreams,
                                 Alternatives alternatives,
                                 std::vector<std::string> const& esproducers,
                                 std::filesystem::path const& datadir,
                                 bool validation)
      : source_(warmupEvents, maxEvents, runForMinutes, registry_, datadir, validation) {
    for (auto const& name : esproducers) {
      pluginManager_.load(name);
      auto esp = ESPluginFactory::create(name, datadir);
      esp->produce(eventSetup_);
    }

    // normalise the total weight to the number of streams
    float total = 0.;
    for (auto const& alternative : alternatives) {
      total += alternative.weight;
    }
    schedules_.reserve(numberOfStreams);
    float cumulative = 0.;
    int lower_range = 0;
    int upper_range = 0;
    for (auto& alternative : alternatives) {
      cumulative += alternative.weight;
      lower_range = upper_range;
      upper_range = static_cast<int>(std::round(cumulative * numberOfStreams / total));
      for (int i = lower_range; i < upper_range; ++i) {
        schedules_.emplace_back(registry_, pluginManager_, &source_, &eventSetup_, i, alternative.path);
      }
      streamsPerBackend_.emplace_back(alternative.backend, upper_range - lower_range);
    }
  }

  void EventProcessor::runToCompletion() {
    // The task that waits for all other work
    FinalWaitingTask globalWaitTask;
    tbb::task_group group;
    for (auto& s : schedules_) {
      s.runToCompletionAsync(WaitingTaskHolder(group, &globalWaitTask));
    }
    group.wait();
    assert(globalWaitTask.done());
    if (globalWaitTask.exceptionPtr()) {
      std::rethrow_exception(*(globalWaitTask.exceptionPtr()));
    }
    stop_.mark();
  }

  void EventProcessor::endJob() {
    // Only on the first stream...
    schedules_[0].endJob();
  }
}  // namespace edm

File: src/julia-serial/Framework/EventProcessor.jl
using .ESPluginFactory

struct EventProcessor

    numberOfStreams::Int
    source::Source
    event_setup::EventSetup
    registry::ProductRegistry
    schedules::Vector{StreamSchedule}

    # Constructor
    function EventProcessor(numOfStreams::Int, path::Vector{String}, esproducers::Vector{String}, datadir::String, validation::Bool, max_events::Int)
        numberOfStreams = numOfStreams
        registry = ProductRegistry()
        source = Source(registry, datadir, validation, max_events)
        # print(source.raw_events)
        event_setup = EventSetup()
        for name in esproducers
            esp = create_plugin(name, datadir)
            produce(esp, event_setup)
        end

        schedules = Vector{StreamSchedule}()
        for i in 1:numberOfStreams
            push!(schedules, StreamSchedule(registry, source, event_setup, i, path))
        end

        new(numOfStreams, source, event_setup, registry, schedules)
    end
end



function run_processor(ev::EventProcessor)
    # tasks = [Dagger.@spawn run_stream(schedule) for schedule in ev.schedules]
    # # print(length(tasks))
    # # @info "Number of tasks created: $(length(tasks))"
    # # @info "Expected number of streams: $(ev.numberOfStreams)"
    # fetch.(tasks)
    # @threads for i in 1:ev.numberOfStreams
    #     run_stream(ev.schedules[i])
    # end
    @threads for i in 1:ev.numberOfStreams
        run_stream(ev.schedules[i])
    end

    # print(tasks)
    @info "All stream schedules have completed."
end

function warm_up(ev::EventProcessor, run_max_events::Int)
    @threads for i in 1:ev.numberOfStreams
        run_stream(ev.schedules[i])
    end
    ev.source.numEvents[] = 1
    ev.source.max_events = run_max_events
end
